src/routes/transactions/mappers/common/transaction-data.mapper.ts

```diff
@@ -108,61 +107,83 @@ export class TransactionDataMapper {
     if (dataDecoded === null || !isArray(dataDecoded.parameters)) return {};
+    const { method, parameters } = dataDecoded;
+    const promises: Promise<(AddressInfo | null)[] | AddressInfo | null>[] = [];
 
-    const addressInfos = await Promise.all(
-      this.getAddressParametersFromDataDecoded(dataDecoded).map((param) =>
-        this.getIfValidAddress(chainId, param),
-      ),
-    );
-
-    return addressInfos.reduce(
-      (addressInfoIndex, addressInfo) =>
-        addressInfo
-          ? { ...addressInfoIndex, [addressInfo.value]: addressInfo }
-          : addressInfoIndex,
-      {},
-    );
-  }
+    if (method === MULTI_SEND_METHOD_NAME) {
+      for (const parameter of parameters) {
+        const { name, valueDecoded } = parameter;
+        if (name === TRANSACTIONS_PARAMETER_NAME && valueDecoded) {
+          promises.push(this._getFromValueDecoded(chainId, valueDecoded));
+        }
+      }
+    } else {
+      for (const parameter of parameters) {
+        if (parameter.type === ADDRESS_PARAMETER_TYPE) {
+          promises.push(this._getIfValid(chainId, parameter.value));
+        }
+      }
+    }
 
-  private getAddressParametersFromDataDecoded({
-    parameters,
-    method,
-  }: DataDecoded): DataDecodedParameter[] {
-    if (!isArray(parameters)) return [];
-    return method === MULTI_SEND_METHOD_NAME
-      ? this.getAddressParametersFromMultiSend(parameters)
-      : parameters.filter((p) => p.type === ADDRESS_PARAMETER_TYPE);
-  }
+    const addressInfos = (await Promise.all(promises))
+      .flat()
+      .filter((i): i is AddressInfo => i !== null);
 
-  private getAddressParametersFromMultiSend(
-    parameters: DataDecodedParameter[] | null,
-  ): DataDecodedParameter[] {
-    if (!isArray(parameters)) return [];
-    return parameters
-      .filter((p) => p.name === TRANSACTIONS_PARAMETER_NAME)
-      .flatMap((p) =>
-        this.getAddressParametersFromValueDecoded(p.valueDecoded),
-      );
+    return Object.fromEntries(addressInfos.map((i) => [i.value, i]));
   }
 
-  private getAddressParametersFromValueDecoded(
+  /**
+   * Gets an array of {@link AddressInfo} for the passed valueDecoded, by iterating
+   * through its operations. For each operation, both its 'to' address and the addresses
+   * contained in its dataDecoded parameters are collected.
+   *
+   * Null values are added to the result array for each invalid value encountered.
+   * @param chainId - chain id to use
+   * @param valueDecoded - valueDecoded to use
+   */
+  private async _getFromValueDecoded(
+    chainId: string,
     valueDecoded: unknown,
-  ): DataDecodedParameter[] {
+  ): Promise<(AddressInfo | null)[]> {
     if (!isArray(valueDecoded)) return [];
-    return valueDecoded.flatMap((operation) =>
-      this.getAddressParametersFromDataDecoded(operation.dataDecoded),
-    );
+    const promises: Promise<AddressInfo | null>[] = [];
+
+    for (const operation of valueDecoded) {
```

Comment:
nitpick: use `transaction` instead of `operation`. At this stage this should be considered a Multisend.

Reply:
Changed in [d154939](https://github.com/safe-global/safe-client-gateway-nest/pull/381/commits/d1549397cc4351b05e528e841f2e5a849bc90835)